### ISSUE TITLE
[Python] Allowing custom objective / metric function with "objective" and "metric" parameters

### ISSUE THREAD

**Issue by daanklijn** [microsoft/LightGBM#3244]
## Summary

It would be nice if one could register custom objective and loss functions, so that these can be passed into the LightGBM's  `train` function via the `param` argument. 

## Motivation

I believe that this would be a nice feature as this allows for easier hyperparameter tuning. Currently one would need to write logic that retrieves the custom function based on the name in a config file and pass those into the train function's `feval` and `fobj` arguments. It would be much easier and cleaner if one could just register these custom functions s.t. they can be passed in like the standard objective functions and eval metrics.
## Description

Currently one would do something like this to use custom objective functions and evaluation metrics:

```python
import lightgbm as lgb
from custom import custom_objective, custom_metric

params = {}
gbm = lgb.train(params=params,
                train_set=train_set,
                valid_sets=eval_set,
                fobj=custom_objective,
                feval=custom_metric
                )
```

I would propose to also allow something similar to this:

```python
import lightgbm as lgb
from custom import custom_objective, custom_metric

lgb.register_metric(name="custom_metric", function=custom_metric)
lgb.register_objective(name="custom_objective", function=custom_objective)

params = {
    'objective': 'custom_objective',
    'metric': ['custom_metric'],
}

gbm = lgb.train(params=params,
                train_set=train_set,
                valid_sets=eval_set
                )

```

**Comment by guolinke** [comment]
thanks @daanklijn 
I think this is quite easy to implement. 
@shiyu1994 could you help with this when you have time?

**Comment by shiyu1994** [comment]
> thanks @daanklijn
> I think this is quite easy to implement.
> @shiyu1994 could you help with this when you have time?

Sure.

**Comment by StrikerRUS** [comment]
@daanklijn Will passing callable metric/loss in standard params will suit your needs?
We have something similar for sklearn API and I personally find it more convenient than registering callables.
https://github.com/microsoft/LightGBM/blob/3c0e12dc5cf1bdb941e9845d08ba7b891a935745/examples/python-guide/sklearn_example.py#L65
https://github.com/microsoft/LightGBM/blob/78d31d9ae34c9c8ab2b0f8704c6962c8d3510062/python-package/lightgbm/sklearn.py#L519-L521

**Comment by StrikerRUS** [comment]
Closed in favor of being in #2302. We decided to keep all feature requests in one place.

Welcome to contribute this feature! Please re-open this issue (or post a comment if you are not a topic starter) if you are actively working on implementing this feature.

**Comment by TremaMiguel** [comment]
@StrikerRUS , @jameslamb I'm open to develop this

I will consider starting first with `register_objective` because it is only a `callable` or `None`.

I think a similar logic as with Loggers can be used, that is, first register the metric (example [register_logger](https://github.com/microsoft/LightGBM/blob/0075814f02f898ae4070d848078b9d0d4b510ceb/python-package/lightgbm/basic.py#L48)) and then making the `callable` available through a function (example [_log_warning](https://github.com/microsoft/LightGBM/blob/0075814f02f898ae4070d848078b9d0d4b510ceb/python-package/lightgbm/basic.py#L83)) `_retrieve_objective`, this function either return a `callable` or `None`.

Inside `lgb.train()` if `fobj` is not provided check for a previous register `callable`

```python
fobj = fobj or _retrieve_objective()  # either passed argument or registered callable

if fobj is not None:
    for obj_alias in _ConfigAliases.get("objective"):
        params.pop(obj_alias, None)
    params['objective'] = 'none'
```

what are your thoughts?

**Comment by StrikerRUS** [comment]
@TremaMiguel Sorry,  I'm not 100% sold on registering custom functions. I believe that allowing to pass callables to `objective`/`metric` argument is more user friendly and is already implemented in the sklearn wrapper (https://github.com/microsoft/LightGBM/issues/3244#issuecomment-759067234). Adding one more way to pass custom functions will add a lot of maintenance burden and may confuse users by number of possibilities.

**Comment by jameslamb** [comment]
I agree completely with @StrikerRUS

**Comment by TremaMiguel** [comment]
Agree I think registering custom callables could bring some unnecessary logic. I've found this issue available on #2302, should it be marked as completed or removed from the list?

**Comment by StrikerRUS** [comment]
I believe we can wait for some other opinions. Or this issue can be transformed into feature request for allowing custom callables via `objective`/`metric` parameter for all Python interfaces.

**Comment by shiyu1994** [comment]
I agree with @jameslamb @StrikerRUS. Allowing passing customized function with parameter "objective" or "metric" can also fulfill the need in the motivation of this feature request. I'll change the description in #2302.

**Comment by TremaMiguel** [comment]
@jameslamb I'm out of context, as I understand `fobj` and `feval` already accept custom callables, what's different with @StrikerRUS proposal?

**Comment by jameslamb** [comment]
> what's different with @StrikerRUS proposal

We are saying we'd support a pull request making it possible to do something like the following:

```python
bst = lgb.train(
    ...,
    params = {
        "metric": ["auc", _some_function_i_wrote],
    }
)
```

And the same for `objective`.

Like the code linked in https://github.com/microsoft/LightGBM/issues/3244#issuecomment-759067234.

**Comment by TremaMiguel** [comment]
@jameslamb @StrikerRUS so basically to achieve 

```python
bst = lgb.train(
    ...,
    params = {
        "objective": "regression" # or custom function  "objective":custom_objective_function
        "metric": ["auc", _some_function_i_wrote],
    }
)
```
one needs to get the following done

1. Check if `objective` is in params and assigned it to `fobj` like the [R implementation](https://github.com/microsoft/LightGBM/blob/master/R-package/R/lgb.train.R#L107). This will be passed to [Booster.update()](https://github.com/microsoft/LightGBM/blob/master/python-package/lightgbm/engine.py#L236)
2. Check if `metric` is in params and pass it to [Booster.eval_train()](https://github.com/microsoft/LightGBM/blob/master/python-package/lightgbm/engine.py#L242) and [Booster.eval_valid()](https://github.com/microsoft/LightGBM/blob/master/python-package/lightgbm/engine.py#L243)
3. Add support for multiple `metric` values like the [R implementation](https://github.com/microsoft/LightGBM/blob/master/R-package/R/lgb.train.R#L301), creating a list of [eval_functions](https://github.com/microsoft/LightGBM/blob/master/R-package/R/lgb.train.R#L114)

is there anything missing?

**Comment by StrikerRUS** [comment]
Firstly I guess we can simply transfer this logic 
https://github.com/microsoft/LightGBM/blob/01568cf59a412de7823b373fe13eb36db9cffe53/python-package/lightgbm/sklearn.py#L668-L675
into `basic.py`

**Comment by TremaMiguel** [comment]
@StrikerRUS Ohh I see, I think this one is kind of easy, one just need to handle:

* `metrics` with the [code](https://github.com/microsoft/LightGBM/blob/01568cf59a412de7823b373fe13eb36db9cffe53/python-package/lightgbm/sklearn.py#L668-L675) you provide, ignoring `feval` if params has a `metric` key.
* handle `fobj` if `callable` or `string` with

```python
# objective is callable
f_obj = params['objective']
if callable(f_obj):
    for obj_alias in _ConfigAliases.get("objective"):
        params.pop(obj_alias, None)
    fobj = params['objective']
    params['objective'] = None
elif isinstance(f_obj, str):
    for obj_alias in _ConfigAliases.get("objective"):
        params.pop(obj_alias, None)
```

and no need to add a list of `eval_functions` cause [Booster.__inner_eval](https://github.com/microsoft/LightGBM/blob/master/python-package/lightgbm/basic.py#L3831) supports it.
